examples/misc/navigate_search_result_pages_caching_tokens.rb

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env ruby
+# Encoding: utf-8
+#
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# GoogleAdsService.Search results are paginated but they can only be retrieved
+# in sequence starting by the first page. More details at
+# https://developers.google.com/google-ads/api/docs/reporting/paging.
+#
+# This example searches campaigns illustrating how GoogleAdsService.Search
+# result page tokens can be cached and reused to retrieve previous pages. This
+# is useful when you need to request pages that were already requested in the
+# past without starting over from the first page. For example, it can be used
+# to implement an interactive application that displays a page of results at a
+# time without caching all the results first.
+#
+# To add campaigns, run add_campaigns.rb.
+
+require 'optparse'
+require 'google/ads/google_ads'
+
+def navigate_search_result_pages_caching_tokens(customer_id)
+  # GoogleAdsClient will read a config file from
+  # ENV['HOME']/google_ads_config.rb when called without parameters
+  client = Google::Ads::GoogleAds::GoogleAdsClient.new
+
+  # The cache of page tokens. The key is the 0-indexed page number.
+  # The first page's token is always the empty string.
+  page_tokens = {
+    0 => ''
+  }
+
+  puts "Fetching page 1 to get metadata"
+  query = "SELECT campaign.id, campaign.name FROM campaign ORDER BY campaign.name LIMIT #{RESULTS_LIMIT}"
+  response = client.service.google_ads.search(
+    customer_id: customer_id,
+    query: query,
+    page_size: PAGE_SIZE,
+    return_total_results_count: true,
+  )
+
+  cache_next_page_token(page_tokens, response.page, 1)
+
+  # The total results count does not take into consideration the LIMIT clause
+  # of the query so we need to find the minimal value between the limit and the
+  # total results count.
+  total_results = [
+    RESULTS_LIMIT,
+    response.page.response.total_results_count,
+  ].min
+  puts "Found ${total_results} campaigns."
+
+  total_pages = (total_results.to_f / PAGE_SIZE).ceil
+  raise "Could not find any campaigns." if total_pages == 0
+
+  middle_page = (total_pages / 2.0).ceil
+  puts "Printing results from page #{middle_page}."
+  fetch_and_print_page_results(
+    client,
+    customer_id,
+    query,
+    PAGE_SIZE,
+    true,
+    middle_page,
+    page_tokens,
+  )
+
+  puts 'Print all pages, starting at the last page and moving to the first.'
+  (1..total_pages).reverse_each do |page_number|
```

Comment:
Should that `1` be a `0`? Otherwise you won't print the first page.

Reply:
"page_tokens" is 0-indexed, cache_next_page_token expects a 0-indexed page number, and fetch_and_print_page_results expects a 1-indexed page number.

It all works the way it's written, but it is a little confusing. This is following the design in the golden.

Reply:
That seems awfully confusing. If you think Ruby users would be less confused if you made `page_tokens` and its usages 1-indexed, please go ahead and make that change. I suspect the PHP golden did it this way because it was using a 0-based array instead of a map (or chose to use 0-based keys in a map), but I don't think you have to follow that exactly.
@PierrickVoulet , please chime in if I missed something.

Reply:
That's actually a good point. My programmer wanted everything to be 0-indexed but the 1-indexed approach makes more sense here. I updated it.

Reply:
Java implemented the 1-indexed solution, feel free to replicate https://github.com/googleads/google-ads-java/blob/main/google-ads-examples/src/main/java/com/google/ads/googleads/examples/misc/NavigateSearchResultPagesCachingTokens.java#L109.

I agree this can be confusing, I will make the improvement in PHP as well.

Reply:
Here is how the change looks like in PHP: https://github.com/googleads/google-ads-php/pull/777.

Reply:
Nice - much easier to interpret now. Thank you both!